pkg/workloads/spark_job/spark_util.py

```diff
@@ -460,6 +487,37 @@ def validate_transformer(column_name, df, ctx, spark):
     trans_impl, _ = ctx.get_transformer_impl(column_name)
 
     if hasattr(trans_impl, "transform_python"):
+        sample_df = df.collect()
+        sample = sample_df[0]
+        inputs = ctx.create_column_inputs_map(sample, column_name)
+        _, impl_args = extract_inputs(column_name, ctx)
+        transformedSample = trans_impl.transform_python(inputs, impl_args)
+        rowType = type(transformedSample)
+        isList = rowType == list
+
+        for row in sample_df:
+            inputs = ctx.create_column_inputs_map(row, column_name)
+            transformedSample = trans_impl.transform_python(inputs, impl_args)
+            if rowType != type(transformedSample):
+                raise UserRuntimeException(
+                "transformed column " + column_name,
+                "type inference failed, mixed data types in dataframe.",
+            )
+
+
+        typeConversionDict = PYTHON_TYPE_TO_CORTEX_TYPE
+        if isList:
+            rowType = type(transformedSample[0])
+            typeConversionDict = PYTHON_TYPE_TO_CORTEX_LIST_TYPE
+
+        # for downstream operations on this job
+        ctx.transformed_columns[column_name]["type"] = typeConversionDict[rowType]
```

Comment:
Here we set the field in the context, which is inconsistent with how we access these inferred types in other cases. It might be a good "rule" to never modify the context. We can cache the metadata in memory when calling `update_metadata()` (see my comment about this in `context.py`), so later in this job when we call `ctx.get_inferred_colum_type(column_name)` it will already have it cached. Would that work? Which downstream operations use `transformed_columns[column_name]["type"]`, and can we replace calls to it with `ctx.get_inferred_colum_type(column_name)`? I guess this means we should never access `transformed_columns[column_name]["type"]` in Python, since it might be `unknown`.